Шкрылёва Светлана. Технология SEQ-MPI. Минимальное значение элементов вектора. Вариант 4.

SvetlanaShkryleva opened the pull request:
## Описание

- **Задача**: Поиск минимального значения в векторе
- **Вариант**: 4
- **Технология**: SEQ, MPI
- **Описание** 
Параллельный алгоритм с использованием MPI_Scatterv для распределения данных и MPI_Allreduce для нахождения глобального минимума
Последовательная версия на основе std::min_element для сравнения производительности
Поддержка работы с 1, 2, 4, 8 процессами
Полный набор тестов (10 функциональных + производительность)

---

## Чек-лист
<!--
Пожалуйста, убедитесь, что следующие пункты выполнены **до** отправки pull request'а и запроса его ревью:
-->

- [x] **Статус CI**: Все CI-задачи (сборка, тесты, генерация отчёта) успешно проходят на моей ветке в моем форке
- [x] **Директория и именование задачи**: Я создал директорию с именем `<фамилия>_<первая_буква_имени>_<короткое_название_задачи>`
- [x] **Полное описание задачи**: Я предоставил полное описание задачи в теле pull request
- [x] **clang-format**: Мои изменения успешно проходят `clang-format` локально в моем форке (нет ошибок форматирования)
- [x] **clang-tidy**: Мои изменения успешно проходят `clang-tidy` локально в моем форке (нет предупреждений/ошибок)
- [x] **Функциональные тесты**: Все функциональные тесты успешно проходят локально на моей машине
- [x] **Тесты производительности**: Все тесты производительности успешно проходят локально на моей машине
- [x] **Ветка**: Я работаю в ветке, названной точно так же, как директория моей задачи (например, `nesterov_a_vector_sum`), а не в `master`
- [x] **Правдивое содержание**: Я подтверждаю, что все сведения, указанные в этом pull request, являются точными и достоверными

SvetlanaShkryleva committed:
feat: implement parallel vector minimum search using MPI

- Add MPI version with Scatterv/Allreduce for data distribution
- Implement sequential version for performance comparison
- Create comprehensive test suite (10 functional tests + performance tests)
- Add detailed report with experimental analysis
- Support multiple process configurations (1, 2, 4, 8 processes)
- Handle edge cases and data distribution efficiently

All functional tests passed successfully. Performance analysis shows
MPI stability in multi-process environment compared to sequential version.

SvetlanaShkryleva committed:
fix: remove vcpkg submodule configuration

SvetlanaShkryleva committed:
style: apply code formatting fixes

SvetlanaShkryleva committed:
fix: correct MPI_Scatterv usage

SvetlanaShkryleva committed:
fix: all clang-tidy errors

SvetlanaShkryleva committed:
fix: remove const after final in CheckTestOutputData

SvetlanaShkryleva committed:
fix: correct const placement and apply formatting

SvetlanaShkryleva committed:
fix: remove const from CheckTestOutputData in both test files

SvetlanaShkryleva committed:
fix: remove extra bracket and unused includes

SvetlanaShkryleva committed:
fix: clang-tidy signed/unsigned comparison and member initialization

SvetlanaShkryleva committed:
fix: correct MPI logic for finding minimum value

SvetlanaShkryleva committed:
fix: final clang-tidy errors - remove unused includes and redundant initializers

SvetlanaShkryleva committed:
fix clang-tidy warnings: initialize class fields

SvetlanaShkryleva committed:
fix: clang-tidy warnings: use default member initializers

SvetlanaShkryleva committed:
Retry CI tests

SvetlanaShkryleva committed:
fix: correct MPI collective operations for edge cases

SvetlanaShkryleva committed:
fix: set output on all MPI processes

SvetlanaShkryleva committed:
merge master

SvetlanaShkryleva committed:
fix: report

SvetlanaShkryleva committed:
rewrite report according to new documentation requirements

SvetlanaShkryleva committed:
Update report with new performance test data

SvetlanaShkryleva committed:
codecov

SvetlanaShkryleva committed:
clang-tidy

SvetlanaShkryleva committed:
clang-tidy

SvetlanaShkryleva committed:
removed image from data, deleted extra file, reset clang-format to its default configuration

SvetlanaShkryleva committed:
deleted no-lint

SvetlanaShkryleva committed:
clang-tidy

SvetlanaShkryleva committed:
clang-tidy

allnes commented on tasks/shkryleva_s_vec_min_val/mpi/src/ops_mpi.cpp:
```diff
@@ -0,0 +1,94 @@
+  if (GetInput().empty()) {
+    return true;
+  }
+
```
в начале RunImpl стоит `if (GetInput().empty()) return false;`. На не‑нулевых рангах вход по умолчанию пустой, поэтому эти процессы выходят из функции до любых коммуникаций, тогда как ранг 0 пойдет в Scatterv/Bcast и повиснет. Аналогичная проблема в ValidationImpl: там тоже проверяется непустота входа без учета ранга, что приведет к провалу валидации на всех рангах кроме 0. Нужно выполнять проверку только на ранге 0 и распространять результат через MPI_Bcast, а в RunImpl гарантировать участие всех рангов в коммуникациях даже при пустом локальном векторе.

SvetlanaShkryleva committed:
fix uint64_t, Finalize, GetInput().empty()

allnes commented on tasks/shkryleva_s_vec_min_val/mpi/src/ops_mpi.cpp:
```diff
@@ -0,0 +1,94 @@
+  return (GetOutput() == 0);
+}
+
+bool ShkrylevaSVecMinValMPI::PreProcessingImpl() {
```
PreProcessingImpl вызывает collective MPI_Init, PostProcessingImpl дергает MPI_Finalized (при этом финализация закомментирована). Коммуникации и коллективные вызовы согласно требованиям должны находиться в RunImpl; Pre/Post — только локальные проверки или рассылка булева результата с root. Такой перенос нарушает контракт задач, грозит двойной инициализацией/незавершением MPI при множественном запуске задач. Следует убрать MPI_Init/MPI_Finalized из Pre/Post и полагаться на внешнее управление MPI либо перенести необходимые коллективные вызовы в RunImpl.

allnes commented on tasks/shkryleva_s_vec_min_val/mpi/src/ops_mpi.cpp:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  int world_rank = 0;
+  int world_size = 0;
```
В начале RunImpl стоит `if (GetInput().empty()) return false;`. На рангах ≠0 вход никогда не заполняется, поэтому они сразу выходят, не участвуя в `MPI_Bcast`/`MPI_Scatterv`/`MPI_Allreduce`. Это приводит к зависанию/разрушению коммуникаций при корректном запуске с несколькими процессами. Нужно убрать ранний выход и выполнять коллективные вызовы на всех рангах, используя данные только rank 0 и пустые буферы на остальных.

allnes commented on tasks/shkryleva_s_vec_min_val/seq/src/ops_seq.cpp:
```diff
@@ -0,0 +1,46 @@
+    return true;
+  }
+
+  int min_val = GetInput()[0];
```
PostProcessingImpl возвращает false, если минимум равен INT_MIN. Алгоритм корректно находит этот элемент, но из‑за проверки результат помечается как невалидный и тест/задача завершается ошибкой. Аналогичная логика в MPI-версии. Уберите жёсткую проверку `> INT_MIN` (достаточно вернуть true или отдельный флаг корректности) либо храните признак «успешно выполнено» отдельно от значения результата.

SvetlanaShkryleva committed:
fix:
mpi:
- Fix validation to check only on root and broadcast result
- Ensure all ranks participate in collective operations
- Use consistent uint64_t for sizes with proper validation
- Remove MPI calls from Pre/PostProcessing
- Fix scatterv parameters and overflow handling
- Remove INT_MIN restriction in post-processing

seq:
- Fix validation to check input instead of output
- Remove INT_MIN restriction in post-processing
- Clean up redundant empty checks

SvetlanaShkryleva committed:
fix empty vector

SvetlanaShkryleva committed:
fix empty vector in seq

SvetlanaShkryleva committed:
fix empty vector in seq

SvetlanaShkryleva committed:
fix mpi

SvetlanaShkryleva committed:
fix mpi

SvetlanaShkryleva committed:
split into functions

SvetlanaShkryleva committed:
Merge branch 'master' into shkryleva_s_vec_min_val

SvetlanaShkryleva committed:
CI

SvetlanaShkryleva committed:
merged master

SvetlanaShkryleva committed:
CI

SvetlanaShkryleva committed:
deleted example

SvetlanaShkryleva committed:
CI

SvetlanaShkryleva committed:
merged master again

SvetlanaShkryleva committed:
CI without setup environment problems

SvetlanaShkryleva committed:
CI without setup environment problems

SvetlanaShkryleva committed:
Merge branch 'master' into shkryleva_s_vec_min_val

SvetlanaShkryleva committed:
CI without setup environment problems

SvetlanaShkryleva committed:
CI without setup environment problems

SvetlanaShkryleva committed:
Merge branch 'shkryleva_s_vec_min_val' of https://github.com/SvetlanaShkryleva/ppc-2025-processes-engineers into shkryleva_s_vec_min_val
